crates/evm/src/system_calls.rs

```diff
@@ -23,9 +24,113 @@ use revm_primitives::{
     ResultAndState, B256,
 };
 
+/// Apply the [EIP-2935](https://eips.ethereum.org/EIPS/eip-2935) pre block contract call.
+///
+/// This constructs a new [`Evm`] with the given database and environment ([`CfgEnvWithHandlerCfg`]
+/// and [`BlockEnv`]) to execute the pre block contract call.
+///
+/// This uses [`apply_blockhashes_contract_call`] to ultimately apply the blockhash contract state
+/// change.
+#[allow(clippy::too_many_arguments)]
+pub fn pre_block_blockhashes_contract_call<EvmConfig, DB>(
+    db: &mut DB,
+    evm_config: &EvmConfig,
+    chain_spec: &ChainSpec,
+    initialized_cfg: &CfgEnvWithHandlerCfg,
+    initialized_block_env: &BlockEnv,
+    block_number: u64,
+    block_timestamp: u64,
+    parent_block_hash: B256,
+) -> Result<(), BlockExecutionError>
+where
+    DB: Database + DatabaseCommit,
+    DB::Error: Display,
+    EvmConfig: ConfigureEvm,
+{
+    // Apply the pre-block EIP-2935 contract call
+    let mut evm_pre_block = Evm::builder()
+        .with_db(db)
+        .with_env_with_handler_cfg(EnvWithHandlerCfg::new_with_cfg_env(
+            initialized_cfg.clone(),
+            initialized_block_env.clone(),
+            Default::default(),
+        ))
+        .build();
+
+    apply_blockhashes_contract_call(
+        evm_config,
+        chain_spec,
+        block_timestamp,
+        block_number,
+        parent_block_hash,
+        &mut evm_pre_block,
+    )
+}
+
+/// Applies the pre-block call to the [EIP-2935] blockhashes contract, using the given block,
+/// [`ChainSpec`], and EVM.
+///
+/// If Prague is not activated, or the block is the genesis block, then this is a no-op, and no
+/// state changes are made.
+///
+/// [EIP-2935]: https://eips.ethereum.org/EIPS/eip-2935
+#[inline]
+pub fn apply_blockhashes_contract_call<EvmConfig, EXT, DB>(
+    evm_config: &EvmConfig,
+    chain_spec: &ChainSpec,
+    block_timestamp: u64,
+    block_number: u64,
+    parent_block_hash: B256,
+    evm: &mut Evm<'_, EXT, DB>,
+) -> Result<(), BlockExecutionError>
+where
+    DB: Database + DatabaseCommit,
+    DB::Error: core::fmt::Display,
+    EvmConfig: ConfigureEvm,
+{
+    if !chain_spec.is_prague_active_at_timestamp(block_timestamp) {
```

Comment:
do we need to set up the Evm at lines 50-58 above if this is triggered and we'll not use it? other system contract calls are guarded by the higher-level hardfork check, so the `pre_block_blockhashes_contract_call` function isn't even called in case it's not active

Reply:
4788 also includes this check and does the setup, and there is no guarantee on an outer check before calling either of these functions